src/librbd/CopyupRequest.cc

```diff
@@ -0,0 +1,150 @@
+      m_need_copyup(need_copyup), m_parent_completion(NULL),
+      m_copyup_completion(NULL) {}
+
+  CopyupRequest::~CopyupRequest() {
+    assert(m_pending_requests.empty());
+
+    m_ictx->copyup_list_lock.Lock();
+    ldout(m_ictx->cct, 20) << __func__ << " removing the slot " << dendl;
+    map<uint64_t, CopyupRequest*>::iterator it =
+      m_ictx->copyup_list.find(m_object_no);
+    assert(it != m_ictx->copyup_list.end());
+    it->second = NULL;
+    m_ictx->copyup_list.erase(it);
+
+    if (m_ictx->copyup_list.empty())
+      m_ictx->copyup_list_cond.Signal();
+
+    ldout(m_ictx->cct, 20) <<  __func__ << " remove the slot " << m_object_no
+                           << " in copyup_list, size = " << m_ictx->copyup_list.size()
+                           << dendl;
+
+    m_ictx->copyup_list_lock.Unlock();
+  }
+
+  void CopyupRequest::set_ready() {
+    m_ready = true;
+  }
+
+  bool CopyupRequest::is_ready() {
+    return m_ready;
+  }
+
+  bool CopyupRequest::is_need_send_copyup() {
+    return m_need_copyup;
+  }
+
+  ceph::bufferlist& CopyupRequest::get_copyup_data() {
+    return m_copyup_data;
+  }
+
+  Mutex& CopyupRequest::get_lock() {
+    return m_lock;
+  }
+
+  void CopyupRequest::append_request(AioRequest *req) {
+    assert(!m_ready);
+    m_pending_requests.push_back(req);
+  }
+
+  void CopyupRequest::complete_all(int r) {
+    assert(m_ready);
+
+    while (!m_pending_requests.empty()) {
+      vector<AioRequest *>::iterator it = m_pending_requests.begin();
+      AioRequest *req = *it;
+      req->complete(r);
+      m_pending_requests.erase(it);
+    }
+  }
+
+  void CopyupRequest::send_copyup(int r) {
+    ldout(m_ictx->cct, 20) << __func__ << dendl;
+
+    m_ictx->snap_lock.get_read();
+    ::SnapContext snapc = m_ictx->snapc;
+    m_ictx->snap_lock.put_read();
+
+    std::vector<librados::snap_t> snaps;
+    for (std::vector<snapid_t>::const_iterator it = snapc.snaps.begin();
+         it != snapc.snaps.end(); ++it) {
+      snaps.push_back(it->val);
+    }
+
+    librados::ObjectWriteOperation copyup_op;
+    copyup_op.exec("rbd", "copyup", m_copyup_data);
+
+    m_copyup_completion = librados::Rados::aio_create_completion(this, NULL, rbd_copyup_cb);
+    m_ictx->md_ctx.aio_operate(m_oid, m_copyup_completion, &copyup_op,
+                               snapc.seq.val, snaps);
+    m_copyup_completion->release();
+  }
+
+  void CopyupRequest::read_from_parent(vector<pair<uint64_t,uint64_t> >& image_extents)
+  {
+    m_parent_completion = aio_create_completion_internal(this, rbd_read_from_parent_cb);
+    ldout(m_ictx->cct, 20) << __func__ << " this = " << this
+                           << " parent completion " << m_parent_completion
+                           << " extents " << image_extents
+                           << dendl;
+    aio_read(m_ictx->parent, image_extents, NULL, &m_copyup_data, m_parent_completion, 0, true);
+  }
+
+  void rbd_read_from_parent_cb(completion_t cb, void *arg)
```

Comment:
A static member function can still access instance variables (if it has a pointer): 
```
<static> void CopyupRequest::read_from_parent_cb(completion_t cb, void *arg) {
  CopyupRequest *req = reinterpret_cast<CopyupRequest *>(arg);
  req->m_lock.Lock();
  ...
}
```